# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGLContext.h
# Commit: bbfe23c61add
# Full Hash: bbfe23c61add07377daf80696fd6996c0463f5b6
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-06-09 09:21:34
# Regressor Bug: 1632249
# File Overlap Count: 1
# Description:
#   Bug 1632249 - Support compositing out-of-process WebGL. r=handyman,lsalzman,nical,geckoview-reviewers,agi,imanol
#   
#   * Majorly simplity CanvasRenderer
#   * Replace GLScreenBuffer with trivial GLSwapChain
#   * Use descriptor structs so that future SharedSurface changes aren't so painful
# ==============================================================================

diff -r 8fcfac450ecc -r bbfe23c61add dom/canvas/WebGLContext.h
--- a/dom/canvas/WebGLContext.h	Mon Jun 08 20:13:01 2020 +0000
+++ b/dom/canvas/WebGLContext.h	Mon Jun 08 20:34:15 2020 +0000
@@ -11,6 +11,7 @@
 
 #include "GLContextTypes.h"
 #include "GLDefs.h"
+#include "GLScreenBuffer.h"
 #include "mozilla/Attributes.h"
 #include "mozilla/CheckedInt.h"
 #include "mozilla/dom/BindingDeclarations.h"
@@ -268,7 +269,7 @@
     ~LruPosition() { reset(); }
   };
 
-  LruPosition mLruPosition;
+  mutable LruPosition mLruPosition;
 
  public:
   void BumpLru() {
@@ -478,13 +479,14 @@
     return mOptions.preserveDrawingBuffer;
   }
 
-  // Prepare the context for capture before compositing
-  bool PresentScreenBuffer(gl::GLScreenBuffer* const screen = nullptr);
-  bool PresentScreenBufferVR(gl::GLScreenBuffer* const screen = nullptr,
-                             const gl::MozFramebuffer* const fb = nullptr);
+  // Present to compositor
+ private:
+  bool PresentInto(gl::SwapChain& swapChain);
 
-  // Present to compositor
-  bool Present();
+ public:
+  void Present();
+  RefPtr<gfx::DataSourceSurface> GetFrontBufferSnapshot();
+  Maybe<layers::SurfaceDescriptor> GetFrontBuffer(layers::TextureType);
 
   void RunContextLossTimer();
   void CheckForContextLoss();
@@ -584,10 +586,6 @@
   void PixelStorei(GLenum pname, uint32_t param);
   void PolygonOffset(GLfloat factor, GLfloat units);
 
-  RefPtr<layers::SharedSurfaceTextureClient> GetVRFrame(WebGLFramebuffer* fb);
-  void ClearVRFrame();
-  void EnsureVRReady();
-
   ////
 
   webgl::PackingInfo ValidImplementationColorReadPI(
@@ -1178,7 +1176,7 @@
 
   // Used for some hardware (particularly Tegra 2 and 4) that likes to
   // be Flushed while doing hundreds of draw calls.
-  int mDrawCallsSinceLastFlush = 0;
+  mutable uint64_t mDrawCallsSinceLastFlush = 0;
 
   mutable uint64_t mWarningCount = 0;
   const uint64_t mMaxWarnings;
@@ -1202,9 +1200,6 @@
   bool mNeedsIndexValidation = false;
 
   const bool mAllowFBInvalidation;
-#if defined(MOZ_WIDGET_ANDROID)
-  UniquePtr<gl::GLScreenBuffer> mVRScreen;
-#endif
 
   bool Has64BitTimestamps() const;
 
@@ -1216,6 +1211,8 @@
   mutable bool mDefaultFB_IsInvalid = false;
   mutable UniquePtr<gl::MozFramebuffer> mResolvedDefaultFB;
 
+  gl::SwapChain mSwapChain;
+
   // --
 
   bool EnsureDefaultFB();